File: app/src/main/java/com/example/animabuilder/CharacterCreation/Attributes/SecondaryCharacteristic.java
package com.example.animabuilder.CharacterCreation.Attributes;

import com.example.animabuilder.CharacterCreation.SerialOutputStream;

import java.io.BufferedReader;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.Serializable;
import java.nio.charset.StandardCharsets;

public class SecondaryCharacteristic implements Serializable {
    ByteArrayOutputStream byteArray;

    int modVal;
    int pointsApplied;
    int devPerPoint;
    int pointsFromClass;
    boolean bonusApplied;
    int total;

    public int getModVal() {
        return modVal;
    }

    public void setModVal(int value){
        modVal = value;
        refreshTotal();
    }

    public int getPointsApplied() {
        return pointsApplied;
    }

    public void setPointsApplied(int points){
        pointsApplied = points;
        refreshTotal();
    }

    public int getDevPerPoint(){
        return devPerPoint;
    }

    public void setDevPerPoint(int perPoint){
        devPerPoint = perPoint;
    }

    public int getPointsFromClass() {
        return pointsFromClass;
    }

    public void setPointsFromClass(int points){
        pointsFromClass = points;
        refreshTotal();
    }

    public boolean isBonusApplied() {
        return bonusApplied;
    }

    public void setBonusApplied(boolean bonus){
        bonusApplied = bonus;
        refreshTotal();
    }

    public SecondaryCharacteristic(){
        modVal = 0;
        pointsApplied = 0;
        pointsFromClass = 0;
        bonusApplied = false;

        refreshTotal();
    }

    private void refreshTotal(){
        total = modVal + pointsApplied + pointsFromClass;
        if(pointsApplied == 0)
            total -= 30;
        else if(bonusApplied)
            total += 5;
    }

    public int getTotal(){
        return total;
    }

    public void load(BufferedReader fileReader, SecondaryList caller) throws IOException {
        setPointsApplied(Integer.parseInt(fileReader.readLine()));

        String loadBoolean = fileReader.readLine();
        if(loadBoolean.equals("true")) {
            setBonusApplied(true);
            caller.incrementNat(true);
        }
        else
            setBonusApplied(false);
    }

    public void write(SerialOutputStream byteArray){
        byteArray.write((pointsApplied + "\n").getBytes(StandardCharsets.UTF_8), 0, (pointsApplied + "\n").getBytes(StandardCharsets.UTF_8).length);

        String booleanIn;
        if(bonusApplied)
            booleanIn = "true";
        else
            booleanIn = "false";

        byteArray.write((booleanIn + "\n").getBytes(StandardCharsets.UTF_8), 0, (booleanIn + "\n").getBytes(StandardCharsets.UTF_8).length);
    }
}

File: app/src/main/java/com/example/animabuilder/CharacterCreation/Attributes/SecondaryAbilities/SecondaryCharacteristic.kt
package com.example.animabuilder.CharacterCreation.Attributes.SecondaryAbilities

import kotlin.Throws
import com.example.animabuilder.CharacterCreation.Attributes.SecondaryAbilities.SecondaryList
import com.example.animabuilder.Serializables.SerialOutputStream
import java.io.BufferedReader
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.Serializable
import java.nio.charset.StandardCharsets

class SecondaryCharacteristic : Serializable {

    var modVal = 0
    var pointsApplied = 0
    var devPerPoint = 0
    var pointsFromClass = 0
    var bonusApplied = false
    var total = 0

    @JvmName("getModVal1")
    fun getModVal(): Int {
        return modVal
    }

    @JvmName("setModVal1")
    fun setModVal(value: Int) {
        modVal = value
        refreshTotal()
    }

    @JvmName("getPointsApplied1")
    fun getPointsApplied(): Int {
        return pointsApplied
    }

    @JvmName("setPointsApplied1")
    fun setPointsApplied(points: Int) {
        pointsApplied = points
        refreshTotal()
    }

    @JvmName("getPointsFromClass1")
    fun getPointsFromClass(): Int {
        return pointsFromClass
    }

    @JvmName("setPointsFromClass1")
    fun setPointsFromClass(points: Int) {
        pointsFromClass = points
        refreshTotal()
    }

    fun isBonusApplied(): Boolean {
        return bonusApplied
    }

    @JvmName("setBonusApplied1")
    fun setBonusApplied(bonus: Boolean) {
        bonusApplied = bonus
        refreshTotal()
    }

    private fun refreshTotal() {
        total = modVal + pointsApplied + pointsFromClass
        if (pointsApplied == 0) total -= 30 else if (bonusApplied) total += 5
    }

    @Throws(IOException::class)
    fun load(fileReader: BufferedReader, caller: SecondaryList) {
        setPointsApplied(fileReader.readLine().toInt())
        val loadBoolean = fileReader.readLine()
        if (loadBoolean == "true") {
            setBonusApplied(true)
            caller.incrementNat(true)
        } else setBonusApplied(false)
    }

    fun write(byteArray: SerialOutputStream) {
        byteArray.write(
            """$pointsApplied
""".toByteArray(StandardCharsets.UTF_8), 0, """$pointsApplied
""".toByteArray(StandardCharsets.UTF_8).size
        )
        val booleanIn: String
        booleanIn = if (bonusApplied) "true" else "false"
        byteArray.write(
            """$booleanIn
""".toByteArray(StandardCharsets.UTF_8), 0, """$booleanIn
""".toByteArray(StandardCharsets.UTF_8).size
        )
    }

    init {
        refreshTotal()
    }
}